node/sync/src/ping.rs

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) 2019-2025 Provable Inc.
+// This file is part of the snarkOS library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::locators::BlockLocators;
+use snarkos_node_router::Router;
+use snarkvm::prelude::Network;
+
+use parking_lot::Mutex;
+use std::{
+    collections::BTreeMap,
+    net::SocketAddr,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+use tokio::{sync::Notify, time::timeout};
+
+/// Internal state of the ping logic
+///
+/// Essentially, ping keeps an ordered map `next_ping` of time(rs) to peer IPs.
+/// When a new peer connects or a Pong message is received, an entry in next ping is created
+/// for when a peer should next be pinged.
+///
+/// TODO (kaimast): maybe keep track of the last ping too, to not trigger spam detection?
```

Comment:
Good one, though I assume typical honest behaviour will only result in at most2-3 messages in a burst, likely not enough for spam detection, but please double check